feat(trace-view): Add errors to the quick trace

wmak opened the pull request:
- This adds errors to the quick trace component, a node will include up to 5 errors and 5 transactions in the dropdown now
- Changed the `TraceFull` type to just be `QuickTraceEvent` to make it easier to work with in the quickTrace
- TODO: going to return `event_id` directly from the backend later instead of having to change it
- Was going to include the issue.id in the dropdown, but we don't want to make it look like we bring the user to the issue, will likely remove it from the backend later.
- Note: the backend will only return errors with the `trace-view-summary` flag effectively feature flagging these changes.
- Also tried to make the dark mode for this a bit better

![image](https://user-images.githubusercontent.com/4205004/110728107-72020f00-81ea-11eb-91e3-230b970ba9d6.png)
![image](https://user-images.githubusercontent.com/4205004/110728156-87773900-81ea-11eb-9d5b-2b9945365dde.png)
<img width="478" alt="image" src="https://user-images.githubusercontent.com/4205004/110846040-50039d80-8279-11eb-90e1-849e8ff46aa9.png">

wmak committed:
feat(trace-view): Add errors to the quick trace

- This adds errors to the quick trace component, a node will include up
  to 5 errors and 5 transactions in the dropdown now
- Changed the TraceFull type to just be QuickTraceEvent to make it
  easier to work with in the quickTrace
- Going to return `event_id` directly from the backend later instead of
  having to change it
- Was going to include the issue.id in the dropdown, but we don't want
  to make it look like we bring the user to the issue

wmak committed:
ref: Make dark mode a little better

wmak commented:
Gonna move to draft first, realized I'll have to update the backend first regardless since an error can happen in a separate project than the transaction.

wmak committed:
Merge branch 'master' into feat/quick-trace-errors

wmak committed:
ref: Adding TraceFull back in and addressing TODO

wmak committed:
ref: revert type changes to traceFullQuery

k-fish reviewed (commented):
More or less looks fine, can look again once it's out of draft. 

Only note is that this dropdown is fairly complex, although since it seems to be organic complexity coming from the use/design I don't know if we can simplify it at all.

wmak commented on src/sentry/static/sentry/app/views/performance/transactionDetails/quickTrace.tsx:
```diff
@@ -347,48 +384,83 @@ function EventNodeSelector({
           );
         })}
+        {nodeKey !== 'current' &&
+          events.slice(0, numEvents).map((event, i) => {
```
One of the main goals of this quick trace is to make navigation between errors + transactions easier which is why they're separate items in the dropdown. So while a transaction will contain the errors, as separate items they make the navigation easier

Zylphrex replied:
I might look into refactoring this component 👀 its gotten a lot more complicated than I initially hoped.

wmak replied:
Mm I removed some copy pasta with 3e5ec7d766584c82d19cb95af9e4e2c8a36d4809 and fixed the Truncate which I broke at some point in this PR.

Zylphrex commented on src/sentry/static/sentry/app/utils/performance/quickTrace/utils.ts:
```diff
@@ -8,7 +8,7 @@ import {DiscoverQueryProps} from 'app/utils/discover/genericDiscoverQuery';
   EventLite,
   QuickTrace,
-  TraceFull,
+  QuickTraceEvent,
```
I believe all the inputs in this file can be kept as `TraceFull` and you'll able to remove all the `?` and it should still work.

Zylphrex commented on src/sentry/static/sentry/app/views/performance/transactionDetails/styles.tsx:
```diff
@@ -52,15 +53,46 @@ export const SectionSubtext = styled('div')`
+
 export const EventNode = styled(Tag)<{pad?: 'left' | 'right'}>`
+  div {
+    color: ${p => nodeColors[p.type || 'white'].color};
+  }
   & ${/* sc-selector */ Background} {
-    border: 1px solid ${p => p.theme.gray500};
+    background-color: ${p => nodeColors[p.type || 'white'].background};
+    border: 1px solid ${p => nodeColors[p.type || 'white'].border};
```
This is fine for now, but we should look into lifting these styles into the `Tag` component.

wmak committed:
ref: Updating types

- Removing children from QuickTraceEvent
- Using TraceFull in utils so we don't need the wonky `?` on children

wmak committed:
ref: Removing some of the copy pasta from quickTrace

- Moving dropdownItem etc. into `DropdownNodeItem` and reusing for both
  errors + events
- Also removeg IssueSubtext and using a prop to dictate colour instead

wmak committed:
ref: Missed a `current?.children`

Zylphrex commented on src/sentry/static/sentry/app/utils/performance/quickTrace/types.ts:
```diff
@@ -19,18 +19,31 @@ export type EventLite = {
+
 export type TraceLite = EventLite[];
 
 /**
  * The `events-trace` endpoint returns a tree structure that gives
  * the parent-child relationships between events.
  */
```
Shouldn't this comment be above `TraceFull`?

Zylphrex commented on src/sentry/static/sentry/app/utils/performance/quickTrace/types.ts:
```diff
@@ -19,18 +19,31 @@ export type EventLite = {
+  errors?: TraceError[];
+};
+
+export type TraceFull = Omit<QuickTraceEvent, 'generation' | 'children' | 'errors'> & {
```
I don't think `'children'` is necessary here since `QuickTraceEvent` does not have a `children` prop. The reason we add `generation` and `errors` here because they exist as optional on `QuickTraceEvent` and we want to make them required.

wmak replied:
Ah yup, missed this when I removed `children` thanks :+1:

wmak committed:
ref: Updating comment + removing omit children
